### ISSUE TITLE
Formatting not working anymore in the Go extension for VS Code

### ISSUE THREAD

**Issue by stevematdavies** [Microsoft/vscode-go#520]
No linting, fmt, import on save etc, nothing supporting go works, since updating to vs .-code 1.5.3 Using go 1.7.1 on Linux Mint 18.

Followed instructions to close gocode, also tried manually and go get also, but entirely broken. Fix not offered.,

**Comment by deejross** [comment]
Unfortunately, this is a recurring theme any time a VSCode update comes out. I typically wait a few weeks before updating to give the plugin author some time to fix the issues that come with new VSCode updates.

**Comment by lukehoban** [comment]
Can anyone else confirm they are seeing this?  Is anyone seeing this on OS/X or Windows? (or only on Linux per @stevematdavies example?)

@stevematdavies If you run the `Developer: Toggle Developer Tools` command, do you see any errors in the console?

**Comment by ironcladlou** [comment]
No issues here with OSX 10.11.6 and Go 1.7.

**Comment by deejross** [comment]
For some reason I thought VSCode 1.5 was the new one, but it's actually 1.6, so I retract my previous comment. One thing I will say is that clearing out all of the go tools in your $GOPATH/bin folder, opening the command pallet, and choosing "Go :Install Tools" did fix a lot of issues for me in VSCode 1.5.3, and now 1.6.

**Comment by ramya-rao-a** [comment]
@stevematdavies Most of the features you are referring to are related to "on save" event.
Can you check if other Go extension features are working for you?
Example:

- `Go: Add Import` command
- autocomplete
- extra info on hover
- etc

**Comment by chrisolsen** [comment]
I am having the same issue with VSCode 1.6.1 and Go 1.6 and I have run Go: Install Tools.

**Comment by ramya-rao-a** [comment]
@chrisolsen Which features do not work for you?

- lint, format and build on save
- autocomplete, 
- extra info on hover, signature help, Goto/Peek definition
- Run tests, coverage

Do you see any info/error message pop up in the VS Code window?

Did this suddenly stop for you after any recent upgrade?

**Comment by chrisolsen** [comment]
I just noticed the issues this morning. I updated to VSCode 1.6.1 last night...I believe

The only issue that I am having is `golint` and `goimports` no longer run on save and none of the settings options are being overridden.

**Comment by ramya-rao-a** [comment]
@chrisolsen Can you share your user/workspace settings?

Looks like the "on save" operations are broken.

Does `goreturns` work for you when you manually format your code? Like `Right-click -> Format Code` or `Alt+Shift+F` ?

What does your output window log when you save?

- Open your "Output" window (View -> Output),
- Select "Go" from the dropdown on the right of the output window
- Make a change in your Go file and save

**Comment by chrisolsen** [comment]
@ramya-rao-a After manually opening the output window, it can be seen that the govet is running. The output window used to auto open up on save, but now does not.

Nothing happens when I select the `Format Code` menu item or use the keyboard shortcut and no action execution is seen in the output window.

```
Finished running tool: C:\Go\bin\go.exe tool vet c:\Users\chris\Go\src\bitbucket.org\quicksort_io\coachchris\account_endpoints.go

Finished running tool: c:\Users\chris\Go\bin\golint.exe
c:\Users\chris\Go\src\bitbucket.org\quicksort_io\coachchris\session.go:49: should not use basic type string as key in context.WithValue

Finished running tool: C:\Go\bin\go.exe build -o C:\Users\chris\AppData\Local\Temp\go-code-check -tags "" .
c:\Users\chris\Go\src\bitbucket.org\quicksort_io\coachchris\account_endpoints.go:106: undefined: log in log.Debug
```

**Comment by ramya-rao-a** [comment]
According to the output, linting is working. You don't see any green/red squiggly lines in your code?

As for formatting, are you seeing any error like "Cannot format due to syntax errors." ?
Any errors in the console after running "Developer: Toggle Developer Tools" in the command pallete and then formatting?

**Comment by chrisolsen** [comment]
I do see the red squiggly lines under the bad code and I can see the yellow lightbulb with suggestions.

There are no console messages shown after running `alt-shift-f`

**Comment by ramya-rao-a** [comment]
So, then listing works for you, that leaves formatting... 
What version of Go extension are you using?
There were changes done to the code around formatting in the latest version, but I don't see how it could silently fail without any user facing error messages.

**Comment by chrisolsen** [comment]
I am using the latest version of the Go extension v0.6.44. From the command line I can run `gofmt`, `goimports` or `goreturns` and each of them work.

**Comment by ramya-rao-a** [comment]
While running `gofmt`, `goimports` or `goreturns`, can you try passing the flag `-d`
One change we did in v0.6.44 in the Go extension is to work with diffs while formatting.

**Comment by chrisolsen** [comment]
Each of the commands returns the following, if that helps.
```
diff account_endpoints.go gofmt/account_endpoints.go
```

**Comment by ramya-rao-a** [comment]
And you see nothing after the line with `diff` command?

I see the same in Windows when the file in question has no applicable changes as per the formatting tool.
Can you add some extra lines/tab spaces to your code, save it (do this outside of VS Code) and then run the same commands in the command line?

**Comment by chrisolsen** [comment]
No, nothing follows the line with the `diff` command. The same is seen if any changes are added inside or outside of vscode and running the same commands. 

It seems that running the command with `gofmt` returns no output. I thought it was the same as `goreturns` and `goimports`, but I could be wrong.

**Comment by ramya-rao-a** [comment]
Did you try adding some extra lines/tab spaces and then trying to format using VS Code?

I have a feeling you are not seeing any changes to your file, because there is no formatting to be done in the first place. That is why nothing follows the `diff` command as there is no difference between your file and the formatted version of your file

**Comment by chrisolsen** [comment]
Yes I tried adding lines :)

Proof is attached
![screenshot 3](https://cloud.githubusercontent.com/assets/41582/19420321/4b6e9afc-93a6-11e6-8e89-9bc1d3a4f841.png)

**Comment by ramya-rao-a** [comment]
Wow, I am stumped, `goreturns` really thinks there is nothing to format. In the same case, can you replace `-d` with `-w` and try?  This would be format the file on disk. If `-w` works for you, but `-d` doesn't then thats something wrong with the formatting tool.

**Comment by chrisolsen** [comment]
Using the `-w` option does perform the formatting and updates the file.

**Comment by ramya-rao-a** [comment]
Ok, then this is a problem with the formatting tool itself. I will release an update that will add a setting to revert back the change to use `-d` in the Go extension. That should unblock you. 

Could you please open an issue with `goreturns` repo for `-d` not working?

**Comment by ramya-rao-a** [comment]
@chrisolsen Could you try getting the latest for `goreturns` and giving `-d` another try? Just in case you have an old version of the tool which had issues with the `-d` flag?

**Comment by chrisolsen** [comment]
I reinstalled it and verified that I now have the latest version of `goreturns`, but I still have the same issue. I will open an issue on it's repo. Thanks for revert 👍

**Comment by chrisolsen** [comment]
Since `goimports -d` also returned the same blank output, would the bug be in the goreturns lib?

**Comment by ramya-rao-a** [comment]
goreturns is based on goimports which is based on gofmt. Do you see the issue while using gofmt too?
And when you tried `goimports -d` did you try using the latest as well?

Logging it in `goimports` would be a good starting point

**Comment by ramya-rao-a** [comment]
@chrisolsen I have published an update to the Go extension : 0.6.45
There is a setting `go.useDiffForFormatting`. Set this to false and you should be unblocked

**Comment by chrisolsen** [comment]
Thank you! Everything is back to normal for me.

**Comment by bezigon** [comment]
You could run the command Go: Install Tools This will install/update all the tools used by the Go extension

**Comment by marpaia** [comment]
Awesome, adding `"go.useDiffForFormatting": false,` to my config fixed this for me as well.

**Comment by ramya-rao-a** [comment]
@marpaia Could you update/upvote https://github.com/sqs/goreturns/issues/28 so that we get some traction on fixing the real issue behind?

**Comment by ramya-rao-a** [comment]
Since this thread took a life of its own, I am renaming the title to "Formatting not working anymore"
@stevematdavies Can you open a separate issue for what you were seeing? We havent heard back from you in a while

**Comment by marpaia** [comment]
@ramya-rao-a done! thanks for shepherding :)

**Comment by ITSecMedia** [comment]
Also trouble on Windows 10 x64 & VSCode 1.7.1 + latest vscode-go 

gofmt -w    doesn't work anymore.
I have to use the Terminal to manually call gofmt on the files to get the job done.

Manual trigger of Go-Format via a popup menu would be help already.

**Comment by ramya-rao-a** [comment]
@ITSecMedia I am replying on the other thread you opened #583

**Comment by ramya-rao-a** [comment]
Closing this thread as there is no action to be taken from the extension's side at the moment.

Anyone else landing here, please try `"go.useDiffForFormatting": false`
If that works, then please update/upvote the upstream issue https://github.com/sqs/goreturns/issues/28 and help the author of `goreturns` debug the issue
